src/main/scala/strawman/collection/mutable/Seq.scala

```diff
@@ -0,0 +1,91 @@
+package strawman.collection.mutable
+
+import java.lang.IndexOutOfBoundsException
+import scala.{Int, Long, Unit, Boolean, Array}
+import strawman.collection
+import strawman.collection.{IterableOnce, toNewSeq, toOldSeq}
+import scala.Predef.intWrapper
+
+trait Seq[A] extends collection.Seq[A]
+                with collection.SeqLike[A, Seq]
+                with Iterable[A] {
+  def update(idx: Int, elem: A): Unit
+  def mapInPlace(f: A => A): this.type
+}
+
+trait GrowableSeq[A] extends Seq[A]
+                        with collection.SeqLike[A, Seq]
+                        with Growable[A] {
+  def insert(idx: Int, elem: A): Unit
+  def insertAll(idx: Int, elems: IterableOnce[A]): Unit
+  def remove(idx: Int): A
+  def remove(from: Int, n: Int): Unit
+  def flatMapInPlace(f: A => IterableOnce[A]): this.type
+  def filterInPlace(p: A => Boolean): this.type
+  def patchInPlace(from: Int, patch: collection.Seq[A], replaced: Int): this.type
+
+  // +=, ++=, clear inherited from Growable
+  def +=:(elem: A): this.type = { insert(0, elem); this }
+  def +=:(elem1: A, elem2: A, elems: A*): this.type = elem1 +=: elem2 +=: elems.toStrawman ++=: this
+  def ++=:(elems: IterableOnce[A]): this.type = { insertAll(0, elems); this }
+
+  def dropInPlace(n: Int): this.type = { remove(0, n); this }
+  def dropRightInPlace(n: Int): this.type = { remove(length - n, n); this }
+  def takeInPlace(n: Int): this.type = { remove(n, length); this }
+  def takeRightInPlace(n: Int): this.type = { remove(0, length - n); this }
+  def sliceInPlace(start: Int, end: Int): this.type = takeInPlace(end).dropInPlace(start)
+
+  def dropWhileInPlace(p: A => Boolean): this.type = {
+    val idx = indexWhere(!p(_))
+    if (idx < 0) { clear(); this } else dropInPlace(idx)
+  }
+  def takeWhileInPlace(p: A => Boolean): this.type = {
+    val idx = indexWhere(!p(_))
+    if (idx < 0) this else takeInPlace(idx)
+  }
+  def padToInPlace(len: Int, elem: A): this.type = {
+    while (length < len) +=(elem)
+    this
+  }
+}
+
+trait IndexedOptimizedSeq[A] extends Seq[A] {
+  def mapInPlace(f: A => A): this.type = {
+    var i = 0
+    while (i < size) { this(i) = f(this(i)); i += 1 }
+    this
+  }
+}
+
+trait IndexedOptimizedGrowableSeq[A] extends IndexedOptimizedSeq[A] with GrowableSeq[A] {
+  def flatMapInPlace(f: A => IterableOnce[A]): this.type = {
+    var i = 0
+    val newElemss = new Array[IterableOnce[A]](size)
```

Comment:
Isn't it better to just copy the current elements so that the iterators can be GCed along the way?  And presumably there's already/will be an efficient `toArray` method which would simplify the implementation also.